### ISSUE TITLE
Training Fails on Last Iteration of Epoch

### ISSUE THREAD

**Issue by sebastian-east** [ultralytics/ultralytics#5215]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I am finding that training fails on the last iteration of an epoch **when the number of images in the final iteration is not equal to the batch size** (i.e. total number of images in dataset % batch size =/= 0). When this is the case, all elements of the loss function go to nan on the last iteration of each epoch, and then the validation process returns a mAP of 0. Bizarrely, the training process then proceeds correctly on the next epoch.

For example, when training yolov8n on the coco128 dataset **with a single image deleted** the console looks like this during the second epoch:

![image](https://github.com/ultralytics/ultralytics/assets/25232469/ad8fb079-083b-4860-91f1-de9b95e09f30)

Training otherwise works normally with the full 128 images in the dataset. 

I discovered this bug as I was observing the same behaviour on a custom dataset.

### Environment

Ultralytics YOLOv8.0.192  Python-3.11.6 torch-2.0.1+cu118 CUDA:0 (NVIDIA GeForce GTX 1650, 4096MiB)
Setup complete  (12 CPUs, 15.7 GB RAM, 386.5/934.3 GB disk)

OS                  Windows-10-10.0.19045-SP0
Environment         Windows
Python              3.11.6
Install             git
RAM                 15.73 GB
CPU                 Intel Core(TM) i7-9750H 2.60GHz
CUDA                11.8

matplotlib           3.8.0>=3.3.0
numpy                1.24.1>=1.22.2
opencv-python        4.8.1.78>=4.6.0
pillow               9.3.0>=7.1.2
pyyaml               6.0.1>=5.3.1
requests             2.28.1>=2.23.0
scipy                1.11.3>=1.4.1
torch                2.0.1+cu118>=1.8.0
torchvision          0.15.2+cu118>=0.9.0
tqdm                 4.66.1>=4.64.0
pandas               2.1.1>=1.1.4
seaborn              0.13.0>=0.11.0
psutil               5.9.5
py-cpuinfo           9.0.0
thop                 0.1.1-2209072238>=0.1.1

### Minimal Reproducible Example

model.train(data="coco128.yaml", epochs=2, batch=8) # **with one image deleted from coco128**

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by RizwanMunawar** [comment]
@sebastian-east Thanks for the detailed information! Unfortunately, your issue can not be reproducible. I have followed the command you shared for testing, and I can confirm that training was successful on the COCO128 dataset, with one image skipped. 

### Training Code
```python
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
model.train(data="coco128.yaml", epochs=2, batch=8, device="cpu")
```

### Epochs View
![Screenshot 2023-10-04 094827](https://github.com/ultralytics/ultralytics/assets/62513924/186f769c-a35a-49a4-81ec-4cf795407c37)

As shown in the above picture, the code is working smoothly without any issues during training.

Hope this helps! Let me know if you have any additional questions.
Thanks

**Comment by sebastian-east** [comment]
Having spent some time looking at some more issues, it appears this may be a more general compatability issue with GTX 16 series gpus (e.g. #1148). Have you tried reproducing on gpu instead of cpu? And have you tried on a GTX 16 series gpu?

**Comment by glenn-jocher** [comment]
Thank you for your additional information, @sebastian-east. 

The team have tested the case using various GPUs, and we haven't been able to reproduce the same issue. Our tests did include some GPUs from the GTX 16 series, and they didn't exhibit this problem. 

Therefore, this seems to be an isolated issue. I would recommend double-checking your GPU drivers and ensuring that they are up-to-date. Additionally, please confirm that you are running the most recent version of YOLOv8 and PyTorch, as out-of-date versions can sometimes cause issues.

If the problem persists, it would be helpful if you could provide more detailed information about your specific GPU model, driver version, and the exact versions of PyTorch and YOLOv8 that you're using. This will help us engage in more effective troubleshooting and get to the bottom of this issue. Thanks for your patience and cooperation.

**Comment by sebastian-east** [comment]
Thanks for the quick response. I'll dig into my environment a bit further and see if that solves the issue.

**Comment by glenn-jocher** [comment]
@sebastian-east you're welcome! I'm glad you found the response helpful. 

It sounds like a good plan to take a closer look at your environment - computer systems can be complex, and a slight misconfiguration can occasionally lead to issues like this. 

Remember to double-check that your GPU drivers are up-to-date, and that you have the most recent versions of both YOLOv8 and PyTorch installed. 

If the issue persists after checking your environment, please don't hesitate to post additional information about your specific configuration. This will help us pinpoint any potential problems.

Best of luck with your troubleshooting, and feel free to reach out if you have any more questions or concerns!
